### PR TITLE
n_heads> 1 and share_cache support for paged update cache

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
#12422

All post commit test: https://github.com/tenstorrent/tt-metal/actions/runs/10962505962

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 1]
```diff
@@ -65,30 +86,53 @@ def run_test_update_cache_decode(
         eq_update, output_update = comp_pcc(x, tt_updated_slice)  # checks the updated parts
     logger.debug(output_cache)
     logger.debug(output_update)
+
+    if not eq_cache or not eq_update:
```
Comment (caixunshiren) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:90]: we should remove this or change it to logger debug, definitely remove the breakpoint

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 2]
```diff
@@ -65,30 +86,53 @@ def run_test_update_cache_decode(
...
+                    logger.info(f"cache at {i}: {cache[:, :, i, :]}")
+                    logger.info(f"tt_got_back at {i}: {tt_got_back[:, :, i, :]}")
+            else:
+                eq, pcc = comp_pcc(cache[:, :, i, :], tt_got_back[:, :, i, :])
+                if not eq:
+                    logger.error(f"cache_delta {pcc} pcc at {i}: {cache_delta[:, :, i, :]}")
+                    logger.info(f"cache at {i}: {cache[:, :, i, :]}")
+                    logger.info(f"tt_got_back at {i}: {tt_got_back[:, :, i, :]}")
+
+        breakpoint()
```
Comment (caixunshiren) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:106]: remove breakpoint

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 3]
```diff
@@ -124,9 +168,19 @@ def test_update_cache_decode(
         sharded_high = ttnn.Tensor(x_pad, input_dtype).to(ttnn.TILE_LAYOUT).to(device, input_mem_config)
         sharded_reserved.deallocate(True)
 
-    run_test_update_cache_decode(
-        cache_idx, head_dim, max_seq_len, num_users, num_heads, input_dtype, cache_dtype, device
-    )
+    for idx in range(10):
```
Comment (caixunshiren) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:171]: might not need to run 10 times

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/compute/update_cache.cpp [hunk 4]
```diff
@@ -7,6 +7,8 @@
 #include "compute_kernel_api/common.h"
 #include "compute_kernel_api/pack_untilize.h"
 #include "compute_kernel_api/tilize.h"
+#include "debug/dprint.h"  // required in all kernels using DPRINT
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/compute/update_cache.cpp:10]: get rid of dprint

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp [hunk 5]
```diff
@@ -17,22 +17,28 @@ void PagedUpdateCacheDeviceOperation::validate(const std::vector<Tensor>& input_
     TT_FATAL(input_tensor.device() == cache_tensor.device(), "Operands to update_cache need to be on the same device!");
     TT_FATAL(input_tensor.buffer() != nullptr and cache_tensor.buffer() != nullptr, "Operands to update_cache need to be allocated in buffers on device!");
     TT_FATAL((input_tensor.get_layout() == Layout::TILE && cache_tensor.get_layout() == Layout::TILE), "Inputs to update_cache must be tilized");
-    TT_FATAL(input_tensor.get_dtype() == DataType::FLOAT32 || input_tensor.get_dtype() == DataType::BFLOAT16 || input_tensor.get_dtype() == DataType::BFLOAT8_B, "Error");
-    TT_FATAL(cache_tensor.get_dtype() == DataType::FLOAT32 || cache_tensor.get_dtype() == DataType::BFLOAT16 || cache_tensor.get_dtype() == DataType::BFLOAT8_B, "Error");
+    TT_FATAL(input_tensor.get_dtype() == DataType::FLOAT32 || input_tensor.get_dtype() == DataType::BFLOAT16, "Data type of input tensor must be FLOAT32 or BFLOAT16");
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp:20]: keep in mind that bfp8 input is disabled due to a correctness issue exposed with multiple heads. We should revisit this with a new issue later

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp [hunk 6]
```diff
@@ -17,22 +17,28 @@ void PagedUpdateCacheDeviceOperation::validate(const std::vector<Tensor>& input_
...
+    TT_FATAL(input_tensor.get_legacy_shape()[-1] == cache_tensor.get_legacy_shape()[-1], "Last dim of input tensor must match last dim of cache tensor");
 
     if (this->op_type == PagedUpdateCacheOpType::UPDATE) {
 
         const bool paged_cache = optional_input_tensors.at(1).has_value();
         uint32_t batch_size;
         if (!paged_cache) {
-            TT_FATAL(cache_tensor.get_legacy_shape()[1] == 1, "Only supports 1 head now.");
-            TT_FATAL(input_tensor.get_legacy_shape()[1] == cache_tensor.get_legacy_shape()[0], "Error");
+            // TT_FATAL(cache_tensor.get_legacy_shape()[1] == 1, "Only supports 1 head now.");
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp:34]: remove the commented code

**Review by caixunshiren — CHANGES_REQUESTED** [review]

**Review by caixunshiren — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp [hunk 7]
```diff
@@ -17,23 +17,28 @@ void PagedUpdateCacheDeviceOperation::validate(const std::vector<Tensor>& input_
...
 
     if (this->op_type == PagedUpdateCacheOpType::UPDATE) {
 
         const bool paged_cache = optional_input_tensors.at(1).has_value();
         uint32_t batch_size;
         if (!paged_cache) {
-            TT_FATAL(cache_tensor.get_legacy_shape()[1] == 1, "Only supports 1 head now.");
-            TT_FATAL(input_tensor.get_legacy_shape()[1] == cache_tensor.get_legacy_shape()[0], "Error");
+            if (this->share_cache){
+                TT_FATAL(cache_tensor.get_legacy_shape()[0] == 1, "Error");
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp:35]: Could we get better error messages?

Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.cpp]: Sounds good. My latest changes should addressed this comment

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 8]
```diff
@@ -65,30 +86,51 @@ def run_test_update_cache_decode(
...
+                if not eq:
+                    logger.error(f"cache_delta {pcc} pcc at {i}: {cache_delta[:, :, i, :]}")
+                    logger.info(f"cache at {i}: {cache[:, :, i, :]}")
+                    logger.info(f"tt_got_back at {i}: {tt_got_back[:, :, i, :]}")
+
     assert eq_cache and eq_update
 
 
-@pytest.mark.skip("Test case covered by others")
+@pytest.mark.skipif(is_grayskull(), reason="Grayskull does not support paged cache")
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:109]: Use `skip_for_grayskull`

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 9]
```diff
@@ -65,30 +86,51 @@ def run_test_update_cache_decode(
...
 @pytest.mark.parametrize("head_dim", [128])
-@pytest.mark.parametrize("max_seq_len", [2048])
-@pytest.mark.parametrize("num_users", [32])
-@pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
-@pytest.mark.parametrize("cache_idx", [0, 1, 127, 1057])
-@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("max_seq_len", [2048, 32 * 1024])
+@pytest.mark.parametrize("num_users", [4])
+@pytest.mark.parametrize("num_heads", [8])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:115]: Test parameters have changed. Do we not care about the old cases anymore?

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 10]
```diff
@@ -149,9 +201,9 @@ def check_zero(tensor):
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
 @pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
 @pytest.mark.parametrize("cache_idx", [127, 1057])
-@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat8_b])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:206]: Same comment about changing the test parameters. Should we keep the old parameters?

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 11]
```diff
@@ -271,9 +323,9 @@ def run_test_tensor_index_update_cache_decode(
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
 @pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
 @pytest.mark.parametrize("cache_idx", [0, 1, 127, 1057])
-@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat8_b])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:328]: Same comment

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 12]
```diff
@@ -295,9 +347,9 @@ def test_tensor_index_update_cache_decode(
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
 @pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
 @pytest.mark.parametrize("cache_idx", [127, 1057])
-@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat8_b])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:352]: Same comment

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 13]
```diff
@@ -424,13 +477,13 @@ def run_test_paged_update_cache_decode(
     assert eq_cache and eq_update
 
 
-@pytest.mark.skip("Test case covered by others")
+# @pytest.mark.skip("Test case covered by others")
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:480]: Remove this if we don't need it anymore.

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 14]
```diff
@@ -424,13 +477,13 @@ def run_test_paged_update_cache_decode(
...
 @pytest.mark.parametrize("block_size", [64, 128], ids=["block64", "block128"])
 @pytest.mark.parametrize("head_dim", [128])
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
-@pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("num_heads", [1, 8])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:486]: Same comment

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 15]
```diff
@@ -456,9 +509,9 @@ def test_paged_update_cache_decode(
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
 @pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
 @pytest.mark.parametrize("cache_idx", [127, 1057])
-@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("cache_dtype", [ttnn.bfloat8_b])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:514]: Same comment

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py [hunk 16]
```diff
@@ -612,7 +665,7 @@ def test_paged_fill_cache(
 @pytest.mark.parametrize("max_seq_len", [2048])
 @pytest.mark.parametrize("num_users", [32])
 @pytest.mark.parametrize("num_heads", [1])
-@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16, ttnn.bfloat8_b])
+@pytest.mark.parametrize("input_dtype", [ttnn.bfloat16])
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:668]: Same comment

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/dataflow/reader_update_cache_interleaved_start_id.cpp [hunk 17]
```diff
@@ -34,6 +35,11 @@ void kernel_main() {
     constexpr uint32_t page_table_is_dram = get_compile_time_arg_val(17) == 1;
     constexpr uint32_t page_table_cb_id = get_compile_time_arg_val(18);
 
+    const uint32_t St = get_compile_time_arg_val(19);
+    uint32_t semaphore_addr   = get_semaphore(get_compile_time_arg_val(20));  // semaphore for reciever
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/dataflow/reader_update_cache_interleaved_start_id.cpp:39]: receiver spelling

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/dataflow/writer_update_cache_interleaved_start_id.cpp [hunk 18]
```diff
@@ -30,6 +33,10 @@ void kernel_main() {
     constexpr uint32_t max_blocks_per_seq = get_compile_time_arg_val(14);
     constexpr uint32_t page_table_cb_id = get_compile_time_arg_val(15);
 
+    constexpr uint32_t St = get_compile_time_arg_val(16);
+    uint32_t semaphore_addr   = get_semaphore(get_compile_time_arg_val(17));  // semaphore for reciever
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/kernels/dataflow/writer_update_cache_interleaved_start_id.cpp:37]: receiver spelling

**Review by TT-BrianLiu — CHANGES_REQUESTED** [review]

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py (hunk 9 above)
Comment (caixunshiren) [reply tests/ttnn/unit_tests/operations/test_paged_update_cache.py:115]: This particular one is okay as it was covered by other test cases

**Review by cfjchu — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Pull request merged**